clouddriver-openstack/src/main/groovy/com/netflix/spinnaker/clouddriver/openstack/deploy/ops/securitygroup/UpsertOpenstackSecurityGroupAtomicOperation.groovy

```diff
@@ -47,12 +57,38 @@ class UpsertOpenstackSecurityGroupAtomicOperation implements AtomicOperation<Voi
   * Task status:
   * curl -X GET -H "Accept: application/json" localhost:7002/task/1
   */
+
   @Override
   Void operate(List priorOutputs) {
     task.updateStatus BASE_PHASE, "Upserting security group ${description.name}..."
 
-    description.credentials.provider.upsertSecurityGroup(description.id, description.name, description.description, description.rules)
+    OpenstackClientProvider provider = description.credentials.provider
+
+    // Try getting existing security group, update if needed
+    SecGroupExtension securityGroup
+    if (StringUtils.isNotEmpty(description.id)) {
+      task.updateStatus BASE_PHASE, "Looking up existing security group with id ${description.id}"
+      securityGroup = provider.getSecurityGroup(description.id)
+      if (securityGroup == null) {
```

Comment:
I think we have some inconsistencies in how to handle null results - in the load balancer registration/deregistration, a provider exception is thrown if null is returned.

Then, in the operation classes, you could just to a try catch over the whole method, catch provider exception, do task logging and throw an operation exception with a detailed message (the provider message will be a bit generic).